### ISSUE TITLE
Use __slots__ to define members

### ISSUE THREAD

**Issue by ravenbluedragon** [microsoft/pyright#180]
In the below example, pyright warns **Cannot access member 'a' on type 'Example'**.
Are we able to use `__slots__` to define the members?
My main uses here are importing json from config files and rest apis.

```python
class Example:
    __slots__ = ('a', 'b', 'c')
    def __init__(self, dict):
        for key in self.__slots__:
            setattr(self, key, dict.get(key))
e = Example({'a': 1, 'b': 2, 'c': 3})
print(e.a, e.b, e.c)
```

**Comment by erictraut** [comment]
In the general case, a static analyzer can't support slots because the expression assigned to this magic value could be dynamic. 

I could add code that handles a few special cases where the expression uses only literal values that are known statically, like in your example above, but I'm not sure it's worth it.

Is there a good reason why you're using `__slots__` here? Why not declare these class members the normal way? Or use NamedTuple or data classes? These all allow you to specify members with their associated types and are therefore much more useful if you're interested in static type analysis. The `__slots__` mechanism is a bad design left over from earlier versions of Python.

**Comment by erictraut** [comment]
Haven't heard back, so I'm going to close this request for now. If others have a user for this feature, let me know.

**Comment by iwoloschin** [comment]
@erictraut sorry to bring up an old bug, but I'm curious about what the best path forward would be for using a library that makes use of slots?

https://github.com/ronf/asyncssh/blob/1f1c26f893fa727c4ba6588a72836f5bae790bc7/asyncssh/process.py#L552

The output of a command run is a `SSHCompletedProcess`. Since it just has `__slots__` I get errors saying things like:
```
Cannot access member "returncode" for type "SSHCompletedProcess"
  Member "returncode" is unknownPylance (reportGeneralTypeIssues)
```
From your comment above this sounds like it is working as designed, but since this is an external library I'm not really sure what the options are. Would it be possible for pyright/pylance to detect the slots and just consider them as `Any`? Or would that be worse than doing nothing at all?

I suppose the best answer is that users of 3rd party libraries should file PRs to fix code from a time before static analyzers? 😄

**Comment by erictraut** [comment]
For any library, my recommendation is to make use of a type stub, which should describe the public interface contract of the library in a way that is designed for a static analyzer. You could create the type stubs and contribute them back to the library author or to the typeshed repo.
